prow/servemux/servemux.go

```diff
@@ -0,0 +1,399 @@
+/*
```

Comment:
The factoring looks much better now :)
Discovering this via godoc is still hard though:
```
$ go doc ./prow/servemux/


package servemux // import "k8s.io/test-infra/prow/servemux"

type Metrics struct{ ... }
    func NewMetrics() *Metrics
type PluginMetrics struct{ ... }
type ServeMuxHandlerOptions struct{ ... }
    func NewFakeServeMuxHandlerOptions() *ServeMuxHandlerOption
```

I think it would be better if we had:
```
type ServeMuxHandlerOptions struct{ ... }

type GitHubEventHandlerServer interface {
 RegisterHandleIssueCommentEvent
 RegisterHandlePullRequestEvent
 ...
}

func New(hmacTokenGenerator func() []byte, log *logrus.Entry, bindAddress string) GitHubEventHandlerServer {

}
```

Also, please rename the package. This is not a generic `servemux`. This is a very specific thing to handle github events. Maybe `githubeventserver`?

Reply:
I think it's better not to have a `New` function since this is going to be created with the options.

Reply:
I disagree. Right now what is called `GitHubEventServerOptions` is the actual implementation of the server and this is super confusing. Please make seperate the `Options` from the actual thing, export an Interface `GitHubEventServer` that holds everything that is currently on the `GitHubEventServerOptions` except for `Bind` and `Validate` and add a `New` that returns this interface.

`hook` for example doesn't need the flags.